### PR TITLE
FIX: Only support a subset of markdown.

### PR DESCRIPTION
**Pull request by tgxworld**
Restrict markdown for chat to a subset of features which must be
explictly enabled. The previous pattern which we relied on required us
to selectively opt out. However, anything new added to core will
automatically be made available until it is opted out in this plugin. If
there is any delay in between, this means that `ChatMessage#cooked` can
differ between messages even though `ChatMessage#cooked_version` remains
the same. By switching to this new pattern, we ensure that the way we
cook a message is always the same given a similar
`ChatMessage#cooked_version`.

This commit also fixes a bug where we were trying to opt out of
Markdown-It rules via the `feaures` options which actually doesn't work.

Depends on https://github.com/discourse/discourse/pull/15471

### CODE REVIEW

File: app/models/chat_message.rb [hunk 1]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
 
     Jobs.enqueue(:process_chat_message, args)
   end
-
-  COOK_FEATURES = {
-    anchor: true,
-    "auto-link": true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:74]: No such rule/feature.

File: app/models/chat_message.rb [hunk 2]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
 
     Jobs.enqueue(:process_chat_message, args)
   end
-
-  COOK_FEATURES = {
-    anchor: true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:73]: Migrated to https://github.com/discourse/discourse-chat/pull/468/files#diff-51e0b19328dd306519dc69e052bdaca6fa7c85233020eca3bb73d0d575026f6dR39

File: app/models/chat_message.rb [hunk 3]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    "auto-link": true,
-    bbcode: true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:75]: No such markdown rule or custom markdown plugin.

File: app/models/chat_message.rb [hunk 4]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    "bbcode-block": true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:76]: Migrated to https://github.com/discourse/discourse-chat/pull/468/files#diff-51e0b19328dd306519dc69e052bdaca6fa7c85233020eca3bb73d0d575026f6dR40-R41

File: app/models/chat_message.rb [hunk 5]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    "bbcode-inline": true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:77]: https://github.com/discourse/discourse-chat/pull/468/files#diff-51e0b19328dd306519dc69e052bdaca6fa7c85233020eca3bb73d0d575026f6dR40-R41

File: app/models/chat_message.rb [hunk 6]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    "bold-italics": true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:78]: No such markdown rule or custom markdown plugin.

File: app/models/chat_message.rb [hunk 7]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    "category-hashtag": true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:79]: Migrated to https://github.com/discourse/discourse-chat/pull/468/files#diff-51e0b19328dd306519dc69e052bdaca6fa7c85233020eca3bb73d0d575026f6dR42

File: app/models/chat_message.rb [hunk 8]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    censored: true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:80]: https://github.com/discourse/discourse-chat/pull/468/files#diff-51e0b19328dd306519dc69e052bdaca6fa7c85233020eca3bb73d0d575026f6dR43

File: app/models/chat_message.rb [hunk 9]
```diff
@@ -68,44 +99,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    checklist: false,
```
Comment (tgxworld) [inline app/models/chat_message.rb:81]: https://github.com/discourse/discourse-chat/pull/468/files#diff-51e0b19328dd306519dc69e052bdaca6fa7c85233020eca3bb73d0d575026f6dR44

File: app/models/chat_message.rb [hunk 10]
```diff
@@ -68,44 +99,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    code: true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:82]: https://github.com/discourse/discourse-chat/pull/468/files#diff-51e0b19328dd306519dc69e052bdaca6fa7c85233020eca3bb73d0d575026f6dR59

File: app/models/chat_message.rb [hunk 11]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    "custom-typographer-replacements": false,
```
Comment (tgxworld) [inline app/models/chat_message.rb:83]: Removed because new pattern requires us to opt in by default.

File: app/models/chat_message.rb [hunk 12]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    "d-wrap": false,
```
Comment (tgxworld) [inline app/models/chat_message.rb:84]: Removed because new pattern requires us to opt in by default.

File: app/models/chat_message.rb [hunk 13]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    details: false,
```
Comment (tgxworld) [inline app/models/chat_message.rb:85]: Removed because new pattern requires us to opt in by default.

File: app/models/chat_message.rb [hunk 14]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    "discourse-local-dates": true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:86]: Removed because new pattern requires us to opt in by default.

File: app/models/chat_message.rb [hunk 15]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    emoji: true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:87]: Migrated to https://github.com/discourse/discourse-chat/pull/468/files#diff-51e0b19328dd306519dc69e052bdaca6fa7c85233020eca3bb73d0d575026f6dR45

File: app/models/chat_message.rb [hunk 16]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    emojiShortcuts: true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:88]: No such markdown rule or custom markdown plugin. This is also controlled via a site setting so I don't know why this exists here.

File: app/models/chat_message.rb [hunk 17]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    html: false,
```
Comment (tgxworld) [inline app/models/chat_message.rb:89]: Not sure why this was added but this had no effect on anything.

https://github.com/discourse/discourse/blob/04c77766506184b3b0daccddc51699d34e4eef9e/app/assets/javascripts/pretty-text/addon/engines/discourse-markdown-it.js#L376

File: app/models/chat_message.rb [hunk 18]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    "html-img": true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:90]: Migrated to https://github.com/discourse/discourse-chat/pull/468/files#diff-51e0b19328dd306519dc69e052bdaca6fa7c85233020eca3bb73d0d575026f6dR46

File: app/models/chat_message.rb [hunk 19]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    "inject-line-number": true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:91]: No longer a valid markdown plugin.

File: app/models/chat_message.rb [hunk 20]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    inlineEmoji: true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:92]: No such markdown rule or custom markdown plugin. This is controlled via a site setting in core.

File: app/models/chat_message.rb [hunk 21]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    linkify: true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:93]: This has no effect on anything in core. 

https://github.com/discourse/discourse/blob/04c77766506184b3b0daccddc51699d34e4eef9e/app/assets/javascripts/pretty-text/addon/engines/discourse-markdown-it.js#L379

File: app/models/chat_message.rb [hunk 22]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    mentions: true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:94]: Migrated to https://github.com/discourse/discourse-chat/pull/468/files#diff-51e0b19328dd306519dc69e052bdaca6fa7c85233020eca3bb73d0d575026f6dR47

File: app/models/chat_message.rb [hunk 23]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    newline: true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:95]: Migrated to https://github.com/discourse/discourse-chat/pull/468/files#diff-51e0b19328dd306519dc69e052bdaca6fa7c85233020eca3bb73d0d575026f6dR57

File: app/models/chat_message.rb [hunk 24]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    onebox: true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:96]: Migrated to https://github.com/discourse/discourse-chat/pull/468/files#diff-51e0b19328dd306519dc69e052bdaca6fa7c85233020eca3bb73d0d575026f6dR48

File: app/models/chat_message.rb [hunk 25]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    paragraph: false,
```
Comment (tgxworld) [inline app/models/chat_message.rb:97]: Removed because new pattern requires us to opt in by default.

File: app/models/chat_message.rb [hunk 26]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    policy: false,
```
Comment (tgxworld) [inline app/models/chat_message.rb:98]: Removed because new pattern requires us to opt in by default.

File: app/models/chat_message.rb [hunk 27]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    poll: false,
```
Comment (tgxworld) [inline app/models/chat_message.rb:99]: Removed because new pattern requires us to opt in by default.

File: app/models/chat_message.rb [hunk 28]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    quote: true,
-    quotes: true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:101]: https://github.com/discourse/discourse-chat/pull/468/files#diff-51e0b19328dd306519dc69e052bdaca6fa7c85233020eca3bb73d0d575026f6dR49

File: app/models/chat_message.rb [hunk 29]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    "inject-line-number": true,
-    inlineEmoji: true,
-    linkify: true,
-    mentions: true,
-    newline: true,
-    onebox: true,
-    paragraph: false,
-    policy: false,
-    poll: false,
-    quote: true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:100]: No such markdown rule or custom markdown plugin.

File: app/models/chat_message.rb [hunk 30]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    quotes: true,
-    "resize-controls": false,
```
Comment (tgxworld) [inline app/models/chat_message.rb:102]: Removed because new pattern requires us to opt in by default.

File: app/models/chat_message.rb [hunk 31]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    table: true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:103]: Migrated to https://github.com/discourse/discourse-chat/pull/468/files#diff-51e0b19328dd306519dc69e052bdaca6fa7c85233020eca3bb73d0d575026f6dR60

File: app/models/chat_message.rb [hunk 32]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    "text-post-process": true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:104]: Migrated to https://github.com/discourse/discourse-chat/pull/468/files#diff-51e0b19328dd306519dc69e052bdaca6fa7c85233020eca3bb73d0d575026f6dR50

File: app/models/chat_message.rb [hunk 33]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    unicodeUsernames: false,
```
Comment (tgxworld) [inline app/models/chat_message.rb:105]: No such markdown rule or custom markdown plugin.

File: app/models/chat_message.rb [hunk 34]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    "upload-protocol": true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:106]: https://github.com/discourse/discourse-chat/pull/468/files#diff-51e0b19328dd306519dc69e052bdaca6fa7c85233020eca3bb73d0d575026f6dR51

File: app/models/chat_message.rb [hunk 35]
```diff
@@ -68,44 +98,6 @@ def rebake!(invalidate_oneboxes: false, priority: nil)
...
-    "watched-words": true,
```
Comment (tgxworld) [inline app/models/chat_message.rb:107]: https://github.com/discourse/discourse-chat/pull/468/files#diff-51e0b19328dd306519dc69e052bdaca6fa7c85233020eca3bb73d0d575026f6dR52

**Comment by tgxworld** [comment]
@eviltrout This is an example of how I intend to use the changes I made to core in https://github.com/discourse/discourse/pull/15471. 

@markvanlan Can you have a look at this as well since git blame says you have the most context regarding this.

File: app/models/chat_message.rb (hunk 2 above)
Comment (tgxworld) [reply app/models/chat_message.rb:73]: Actually should we even support this? Does having anchors in chat make sense?

File: spec/models/chat_message_spec.rb [hunk 36]
```diff
@@ -0,0 +1,131 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+describe ChatMessage do
+  describe '.cook' do
+    it 'does not support HTML tags' do
```
Comment (tgxworld) [inline spec/models/chat_message_spec.rb:7]: Test for markdown rules which we do not want is not meant to be exhaustive but I thought it'll be good to test for some of the basic rules that does not make sense for chat.

**Comment by tgxworld** [comment]
https://github.com/discourse/discourse/pull/15471/files need to be merged first before the tests here will pass.

**Review by markvanlan — APPROVED** [review]

File: spec/models/chat_message_spec.rb [hunk 37]
```diff
@@ -0,0 +1,137 @@
...
+    end
+
+    it 'supports emphasis rule' do
+      cooked = ChatMessage.cook("**bold**")
+
+      expect(cooked).to eq("<p><strong>bold</strong></p>")
+    end
+
+    it 'supports link markdown rule' do
+      qa_comment = Fabricate(:qa_comment, raw: "[test link](https://www.example.com)")
```
Comment (martin-brennan) [inline spec/models/chat_message_spec.rb:69]: Noticed a test failure for this, this should be a chat message, copy/paste issue I am sure :)

**Review by martin-brennan — COMMENTED** [review]
@tgxworld I reran the tests now that the other PR is merged, looks like there are some test failures to fix before merging this one.

**Comment by CvX** [comment]
We were using this md syntax to prevent oneboxing some urls (see e.g. bot messages on dev #random)

```md
<https://github.com/discourse/discourse-chat/pull/468>
```

That no longer works. `<` and `>` are rendered as plain text around a link.

**Comment by CvX** [comment]
Oh, and I believe it also broke automatic replacement of emoticons with emoji, e.g. `:P` was replaced with `😛`

**Pull request merged**